### ISSUE TITLE
[Due for payment 2025-04-02] [Due for payment 2025-03-28] [Due for payment 2025-02-18] [Native Share] Implement Native Share for Android 

### ISSUE THREAD

**Issue by grgia** [Expensify/App#48789]
## Description 
This is a tracking issue for the Android implementation of the Native Share project. This section of the project will be handled separately but merged alongside [the IOS issue](https://github.com/Expensify/App/issues/48788). 

## Project 
https://docs.google.com/document/d/14C1VifxvIXeyLAf0XstbC-wvyAb5SQZ0AOw0-jLRyZs/edit

## Tracking  
https://github.com/Expensify/App/issues/36613

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by JmillsExpensify** [comment]
Going through the process.

**Comment by grgia** [comment]
Bumped for update / steps for unblocking permissions

**Comment by BrtqKr** [comment]
I am focusing on split navigation atm. I'll try to complete the flow whenever I'm done with the nav.

**Comment by BrtqKr** [comment]
We've split the work with @filip-solecki and he's taking care of the android atm. We're implementing both platforms in parallel

**Comment by filip-solecki** [comment]
Hi! Yes, I am working on this issue right now

**Comment by grgia** [comment]
Thanks @filip-solecki and @BrtqKr !

**Comment by grgia** [comment]
Hey, @filip-solecki, do you have a draft PR up for this?

**Comment by filip-solecki** [comment]
Yes, it was created a long time ago. I've updated it and working on it! [Here](https://github.com/software-mansion-labs/expensify-app-fork/tree/share-utils/android-other-filetypes-support) it is

**Comment by grgia** [comment]
@filip-solecki Would you post a quick daily update in this issue 🙏

**Comment by filip-solecki** [comment]
Sure! Right now looks like we have probably everything done on the native side - currently I am working on saving data in Onyx and preparing screens for Sharing the image

**Comment by filip-solecki** [comment]
Hi! Yesterday I was working on Submit page

**Comment by grgia** [comment]
Thanks for the updates @filip-solecki! Do you have a rough ETA for when we'll be able to get the PR for android up?

**Comment by filip-solecki** [comment]
Hi! It should be ready by the end of the next week as there is a little additional native code to write. More precise info in[ ios issue](https://github.com/Expensify/App/issues/48788)

**Comment by grgia** [comment]
@filip-solecki Could you post a current update?

**Comment by filip-solecki** [comment]
Hi @grgia! So currently I am working on adding some additional values to native function returning shared object and I am also trying to merge these changes with current iOS branch which has most of shared JS part done

**Comment by grgia** [comment]
@filip-solecki mind posting a quick update here as well?

**Comment by filip-solecki** [comment]
Hi! @Guccio163  is currently working on the shared file preview and remaining JS-related issues, while I am working on adding support for text sharing.

**Comment by grgia** [comment]
@filip-solecki could you give me an update on when you think we could get a draft PR up for android as well?

**Comment by Guccio163** [comment]
Hi @grgia, how do you think we should implement attachment preview section (visible with the donut below) - should it always crop to the predefined dimensions, or maybe scale to fit the whole attachment? It's quite different from Submit flow, because we have to count pdf and videos in, which display in different dynamics than Image in `AttachmentView` (pdfs can be scrolled and always scale to full width, videos can't be cropped).

<img width="255" alt="Screenshot 2024-11-21 at 08 10 45" src="https://github.com/user-attachments/assets/893a1206-7d42-445f-bae1-db8f7ecf570f">

**Comment by grgia** [comment]
@Expensify/design could you help out with[ this question](https://github.com/Expensify/App/issues/48789#issuecomment-2490290690)

**Comment by dannymcclain** [comment]
Is it possible to open the attachment modal from this share screen? Like if I tap on that donut image, do I get the attachment modal showing the whole image?

**Comment by filip-solecki** [comment]
Sure, Android part is done, right now we are working on polishes for iOS as it will be included in one PR

**Comment by filip-solecki** [comment]
> Is it possible to open the attachment modal from this share screen? Like if I tap on that donut image, do I get the attachment modal showing the whole image?

Sure it is

**Comment by Guccio163** [comment]
I'll look closely into it, but from what I've checked it's not that straightforward, because Submit allows only images and right now the biggest problem is making it unified for images, pdfs and videos. Because we are thinking about allowing user to open fullscreen AttachmentView after clicking on the miniature, last friday I came up with an idea to not use rescaled `AttachmentView` in a thumbnail (which is tricky), but rather try to scale thumbnails used in chats for these types of media.
<img width="355" alt="Screenshot 2024-11-26 at 13 35 42" src="https://github.com/user-attachments/assets/dd670cad-bc3b-4c4b-85c8-4ff6a417b6bb">
WDYT @shawnborton ?

**Comment by shawnborton** [comment]
> but rather try to scale thumbnails used in chats for these types of media.

That works for me, I agree we need to support multiple media formats here on the share page so however you think is best to accomplish that works for me.

**Comment by dannymcclain** [comment]
Yup that works for me too! That's kinda where I was going when I asked about the attachment modal. I think since you can "open" the attachment, it's much less important how we display the thumbnail. 

>however you think is best to accomplish that works for me

Agree with this ☝️

**Comment by Guccio163** [comment]
Hi, I'm coming with an update on this issue!

We are successfully displaying thumbnails for `Images` and `Videos`, though I'm not sure if we should align them to center of left, what do you think? 

<img width="475" alt="Screenshot 2024-12-03 at 14 39 31" src="https://github.com/user-attachments/assets/4840c32c-3e1c-4d5c-9e28-7930e7a0df42">

Also, we're using `DefaultAttachmentView` for any file other than image, video and pdf:
 
<img width="340" alt="Screenshot 2024-12-03 at 14 30 42" src="https://github.com/user-attachments/assets/3b3ba480-a27b-4d76-96aa-9346217e6bc9">

That leaves us with PDFs: When PDF thumbnails are being displayed f.ex. in chats, they use .jpg thumbnail images generated specifically on backend. When using share-extension, whole data is sourced from files locally on device, so the only way to properly show PDFs thumbnail is to:

 1. extract first page of the file on device (platform specific)
 2. convert it into .jpg file
 3 a. save file and send it's localisation 
 or
 3 b. send file as a blob
 
Both ways are potentially time consuming, so right now @filip-solecki and I propose to use `DefaultAttachmentView` for PDF just like for all other files and after user clicks on it, show the file in `AttachmentModal` (or not). That of course is available, because `AttachmentModal` uses whole files to display them, but could be misleading to users that some files shown as `DefaultAttachmentView` are able to be viewed in modal and some aren't.

Let me know what you think; We still have to implement extracting video's `aspectRatio` on Android, but I feel we are really close to creating the draft soon 👌

**Comment by shawnborton** [comment]
I think left-aligned makes the most sense, cc @Expensify/design for thoughts though!

**Comment by dannymcclain** [comment]
I'm down with left-aligned.

And re: PDFs, I think it's fine to display them using the "other file" preview style.

>could be misleading to users that some files shown as DefaultAttachmentView are able to be viewed in modal and some aren't

While this is totally true, I personally don't think it's a huge deal. I don't think clicking on a PDF or text file to look at it is going to be common here, so I personally don't think it will lead to much, if any, confusion. If we're REALLY concerned about it, I even think it would be ok to just disable the AttachmentModal for PDFs in this situation. Just my 2 cents!

**Comment by dubielzyk-expensify** [comment]
Left aligned works for me too 👍

**Comment by garrettmknight** [comment]
@grgia we done here?

**Comment by JmillsExpensify** [comment]
I don't believe this issue is being actively used so closing it out. Please re-open if I'm mistaken.

**Comment by garrettmknight** [comment]
@garrettmknight will[ this issue](https://github.com/Expensify/App/issues/58841#issuecomment-2746346438) be fixed as part of the main project or should we keep it separate?

**Comment by rohit9625** [comment]
You mistakenly mentioned yourself @garrettmknight

**Comment by grgia** [comment]
@garrettmknight should be part of this project

**Comment by rohit9625** [comment]
What about the other issues related to Share feature that are open right now?

**Comment by grgia** [comment]
We will need to pay out C+ review to @shubham1206agra for Native Share review  @JmillsExpensify

**Comment by JmillsExpensify** [comment]
Payment summary: @shubham1206agra via Upwork. 

@shubham1206agra I believe you're still on a payment freeze? If so, I'm going to close this issue though please keep it in your records and we can issue payment when you're ready.
